### ISSUE TITLE
How to provide sparsity of the Jacobian to trust-constr solver

### ISSUE THREAD

**Issue by Framarc93** [scipy/scipy#10165]
Hi, I'm trying to solve a simple optimization problem, taken from http://apmonitor.com/che263/index.php/Main/PythonOptimization and modified in order to use non linear constraints. I am trying to provide the sparsity of the jacobian matrix of the constraints as well, to speed up the process, but I have the following error:
```
Traceback (most recent call last):
  File "/usr/lib/python3.6/code.py", line 91, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "/snap/pycharm-professional/132/helpers/pydev/_pydev_bundle/pydev_umd.py", line 197, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "/snap/pycharm-professional/132/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home/francesco/git_workspace/FESTIP_Work/MultipleShooting_Algorithm/testMS5.py", line 48, in <module>
    solution = minimize(objective,x0,method='trust-constr', bounds=bnds,constraints=[conseq], options={'verbose':3})
  File "/usr/local/lib/python3.6/dist-packages/scipy/optimize/_minimize.py", line 613, in minimize
    callback=callback, **options)
  File "/usr/local/lib/python3.6/dist-packages/scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py", line 514, in _minimize_trustregion_constr
    factorization_method)
  File "/usr/local/lib/python3.6/dist-packages/scipy/optimize/_trustregion_constr/tr_interior_point.py", line 309, in tr_interior_point
    constr_eq0, jac_eq0)
  File "/usr/local/lib/python3.6/dist-packages/scipy/optimize/_trustregion_constr/tr_interior_point.py", line 53, in __init__
    self.fun0 = self._compute_function(fun0, constr_ineq0, s0)
  File "/usr/local/lib/python3.6/dist-packages/scipy/optimize/_trustregion_constr/tr_interior_point.py", line 93, in _compute_function
    s[self.enforce_feasibility] = -c_ineq[self.enforce_feasibility]
IndexError: arrays used as indices must be of integer (or boolean) type
```
This is the code I'm using:
```
import numpy as np
from scipy.optimize import minimize, NonlinearConstraint
from scipy.sparse import coo_matrix

def objective(x):
    return x[0]*x[3]*(x[0]+x[1]+x[2])+x[2]

def constraint3(x):
    return np.array(((x[0]**2+x[1]**2 - 4),
                     (x[0]+x[1]-6),
                     (x[2]*x[3]**2-9),
                     (x[2] ** 2 - x[3] - 5)))

n = 4

row  = np.array([0, 0, 1, 1, 2, 2, 3, 3], dtype=np.int8)
col  = np.array([0, 1, 0, 1, 2, 3, 2, 3], dtype=np.int8)
data = np.ones((8), dtype=np.int8)
sparseJac = coo_matrix((data, (row, col)), shape=(4, 4)).toarray()

conseq = NonlinearConstraint(constraint3, (-np.inf, -np.inf, -np.inf, -np.inf), (np.inf, np.inf, np.inf, np.inf), finite_diff_jac_sparsity=sparseJac)
x0 = np.zeros(n)
x0[0] = 1.0
x0[1] = 5.0
x0[2] = 5.0
x0[3] = 1.0

print('Initial Objective: ' + str(objective(x0)))

b = (1.0,5.0)
bnds = (b, b, b, b)

solution = minimize(objective,x0,method='trust-constr', bounds=bnds,constraints=[conseq], options={'verbose':3})
x = solution.x

print('Final Objective: ' + str(objective(x)))

print('Solution')
print('x1 = ' + str(x[0]))
print('x2 = ' + str(x[1]))
print('x3 = ' + str(x[2]))
print('x4 = ' + str(x[3]))
````
What could be the issue? Am I using the NonlinearConstraint function in the wrong way?

Thank you for you time!

**Comment by nmayorov** [comment]
Hi!

First of all, technically github issues is not a QA resource like stackoverflow. But I'm personally fine with answering such questions (and sometimes they might reveal issues with the code).

So for your example:

1. I don't see how the coded constraints correspond to constraints in the problem in the link you gave. How did you come up with them?
2. Putting all bounds to (-np.inf, np.inf) for constraints actually disables them.
3. For such a small constraints amount you don't need to provide Jacobian sparsity.
4. And finally your code example works for me, giving:
```
`gtol` termination condition is satisfied.
Number of iterations: 29, function evaluations: 130, CG iterations: 40, optimality: 6.68e-09, constraint violation: 0.00e+00, execution time: 0.33 s.
Final Objective: 4.003200437286136
Solution
x1 = 1.000207134365899
x2 = 1.0007990157056421
x3 = 1.000399272425455
x4 = 1.0002578418264054
```

Try to update numpy and scipy.

If still doesn't work, provide your versions info, by 
```
python -c "import numpy; import scipy; import sys; print(numpy.__version__, scipy.__version__, sys.version_info)"
```

**Comment by Framarc93** [comment]
Hi, thank you for the answer and sorry for the misuse of github issue section!
 
The constraints are different from the problem I linked because I added them in order to have non linear constraints and they are completely random. I used this code only to test the NonlinearConstraint function and yes, I know that to provide the jacobian for such problem is eccessive.

Anyway, thank you for pointing out that (-np.inf, np.inf) actually disable the constraints. I found that this was the issues so after changing them to one-sided constraints, it works.
